From 1393479c3b19ec41014f07eb78e7cc15f52c9ccb Mon Sep 17 00:00:00 2001
From: Stuart Morgan <smorgan@mythtv.org>
Date: Sat, 2 Aug 2014 12:38:12 +0100
Subject: [PATCH] UPnP: Construct the supported protocolInfo lists from the
 list of support mimetypes

This reduces the number of places that we need to maintain the list of
supported formats.

TODO Instead of hard-coding http-get, have it pull in a list of
supported protocols as well.
---
 mythtv/libs/libmythupnp/httprequest.cpp        |   16 +++++++++
 mythtv/libs/libmythupnp/httprequest.h          |    1 +
 mythtv/libs/libmythupnp/upnputil.cpp           |   43 ++++++++++++++++++++++++
 mythtv/libs/libmythupnp/upnputil.h             |    4 +++
 mythtv/programs/mythbackend/mediaserver.cpp    |   10 +-----
 mythtv/programs/mythfrontend/mediarenderer.cpp |   11 ++----
 6 files changed, 67 insertions(+), 18 deletions(-)

diff --git a/mythtv/libs/libmythupnp/httprequest.cpp b/mythtv/libs/libmythupnp/httprequest.cpp
index f9db0a5..9f4fb8e 100644
--- a/mythtv/libs/libmythupnp/httprequest.cpp
+++ b/mythtv/libs/libmythupnp/httprequest.cpp
@@ -1025,6 +1025,22 @@ QString HTTPRequest::GetMimeType( const QString &sFileExtension )
 //
 /////////////////////////////////////////////////////////////////////////////
 
+QStringList HTTPRequest::GetSupportedMimeTypes()
+{
+    QStringList mimeTypes;
+
+    for (int i = 0; i < g_nMIMELength; i++)
+    {
+        mimeTypes.append( g_MIMETypes[i].pszType );
+    }
+
+    return mimeTypes;
+}
+
+/////////////////////////////////////////////////////////////////////////////
+//
+/////////////////////////////////////////////////////////////////////////////
+
 QString HTTPRequest::TestMimeType( const QString &sFileName )
 {
     QFileInfo info( sFileName );
diff --git a/mythtv/libs/libmythupnp/httprequest.h b/mythtv/libs/libmythupnp/httprequest.h
index 5dc48f6..3ae6919 100644
--- a/mythtv/libs/libmythupnp/httprequest.h
+++ b/mythtv/libs/libmythupnp/httprequest.h
@@ -199,6 +199,7 @@ class UPNP_PUBLIC HTTPRequest
         QString         GetResponseProtocol () const;
 
         static QString  GetMimeType     ( const QString &sFileExtension );
+        static QStringList GetSupportedMimeTypes ();
         static QString  TestMimeType    ( const QString &sFileName );
         static long     GetParameters   ( QString  sParams, QStringMap &mapParams );
         static QString  Encode          ( const QString &sIn );
diff --git a/mythtv/libs/libmythupnp/upnputil.cpp b/mythtv/libs/libmythupnp/upnputil.cpp
index bd781ef..525ea0a 100644
--- a/mythtv/libs/libmythupnp/upnputil.cpp
+++ b/mythtv/libs/libmythupnp/upnputil.cpp
@@ -17,6 +17,7 @@
 
 // Qt headers
 #include <QUuid>
+#include <QStringList>
 
 // MythTV headers
 #include "upnputil.h"
@@ -24,6 +25,7 @@
 #include "compat.h"
 #include "mythconfig.h" // for HAVE_GETIFADDRS
 #include "mythlogging.h"
+#include "httprequest.h"
 
 // POSIX headers 2, needs to be after compat.h for OS X
 #ifndef _WIN32
@@ -314,3 +316,44 @@ QByteArray gzipCompress( const QByteArray &data )
 
     return result;
 }
+
+/**
+ * \brief Return a QStringList containing the supported Source Protocols
+ *
+ * TODO Extend this to dynamically list all supported protocols (e.g. RTSP)
+ *      and any DLNA profile stuff that we can figure out
+ */
+QStringList GetSourceProtocolInfos()
+{
+    QStringList mimeTypes = HTTPRequest::GetSupportedMimeTypes();
+
+    QStringList protocolList;
+    QStringList::Iterator it;
+    for (it = mimeTypes.begin(); it < mimeTypes.end(); ++it)
+    {
+        protocolList << QString("http-get:*:%1:*").arg(*it);
+    }
+
+    return protocolList;
+}
+
+/**
+ * \brief Return a QStringList containing the supported Sink Protocols
+ *
+ * TODO Extend this to dynamically list all supported protocols (e.g. RTSP)
+ *      and any DLNA profile stuff that we can figure out
+ */
+QStringList GetSinkProtocolInfos()
+{
+    QStringList mimeTypes = HTTPRequest::GetSupportedMimeTypes();
+
+    QStringList protocolList;
+    QStringList::Iterator it;
+    for (it = mimeTypes.begin(); it < mimeTypes.end(); ++it)
+    {
+        protocolList << QString("http-get:*:%1:*").arg(*it);
+    }
+
+    return protocolList;
+}
+
diff --git a/mythtv/libs/libmythupnp/upnputil.h b/mythtv/libs/libmythupnp/upnputil.h
index a6524f2..df9ccc3 100644
--- a/mythtv/libs/libmythupnp/upnputil.h
+++ b/mythtv/libs/libmythupnp/upnputil.h
@@ -16,6 +16,7 @@
 #include <QStringList>
 #include <QMap>
 
+#include "upnpexp.h"
 #include "compat.h"     // for suseconds_t
 
 /////////////////////////////////////////////////////////////////////////////
@@ -144,4 +145,7 @@ void AddSecondsToTaskTime ( TaskTime &t, long nSecs );
 
 QByteArray gzipCompress( const QByteArray &data );
 
+UPNP_PUBLIC QStringList GetSourceProtocolInfos ();
+UPNP_PUBLIC QStringList GetSinkProtocolInfos ();
+
 #endif
diff --git a/mythtv/programs/mythbackend/mediaserver.cpp b/mythtv/programs/mythbackend/mediaserver.cpp
index 416f160..5eff338 100644
--- a/mythtv/programs/mythbackend/mediaserver.cpp
+++ b/mythtv/programs/mythbackend/mediaserver.cpp
@@ -200,15 +200,7 @@ void MediaServer::Init(bool bIsMaster, bool bDisableUPnp /* = false */)
 
         if (bIsMaster)
         {
-            QString sSourceProtocols = "http-get:*:image/gif:*,"
-                                       "http-get:*:image/jpeg:*,"
-                                       "http-get:*:image/png:*,"
-                                       "http-get:*:video/avi:*,"
-                                       "http-get:*:audio/mpeg:*,"
-                                       "http-get:*:audio/wav:*,"
-                                       "http-get:*:video/mpeg:*,"
-                                       "http-get:*:video/nupplevideo:*,"
-                                       "http-get:*:video/x-ms-wmv:*";
+            QString sSourceProtocols = GetSourceProtocolInfos().join(",");
 
             LOG(VB_UPNP, LOG_INFO, "MediaServer: Registering MS_MediaReceiverRegistrar Service.");
 
diff --git a/mythtv/programs/mythfrontend/mediarenderer.cpp b/mythtv/programs/mythfrontend/mediarenderer.cpp
index 3d13f6e..3636dda 100644
--- a/mythtv/programs/mythfrontend/mediarenderer.cpp
+++ b/mythtv/programs/mythfrontend/mediarenderer.cpp
@@ -258,15 +258,8 @@ MediaRenderer::MediaRenderer(): m_pUPnpCMGR(NULL)
         m_pHttpServer->RegisterExtension(
             new MythFrontendStatus(m_pHttpServer->GetSharePath()));
 
-        QString sSinkProtocols = "http-get:*:image/gif:*,"
-                                 "http-get:*:image/jpeg:*,"
-                                 "http-get:*:image/png:*,"
-                                 "http-get:*:video/avi:*,"
-                                 "http-get:*:audio/mpeg:*,"
-                                 "http-get:*:audio/wav:*,"
-                                 "http-get:*:video/mpeg:*,"
-                                 "http-get:*:video/nupplevideo:*,"
-                                 "http-get:*:video/x-ms-wmv:*";
+        QString sSinkProtocols = GetSinkProtocolInfos().join(",");
+
         // ------------------------------------------------------------------
         // Register the MythFEXML protocol...
         // ------------------------------------------------------------------
-- 
1.7.10.2

